File: web/calc_client.py
import sys
import logging
from pathlib import Path
import requests

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from core.logger import make_logger

class CalcClient:
    def __init__(self, base_url='http://127.0.0.1:8000', logger=None):
        self.base_url = base_url
        if logger is None:
            logger_name = self.__class__.__name__
            self.logger = make_logger(name=logger_name, file=False, console=False, level=logging.WARNING)
        else:
            self.logger = logger

        self.data = self.get("get_state", params = {"full": True})
        self.logger.info("CalcClient connected and loaded data %s", list(self.data))
    
    # Запросы
    def get(self, endpoint, params=None):
        if params is None:
            params = {}
        try:
            response = requests.get(
                    f"{self.base_url}/calc/{endpoint}",
                    params=params,
                    timeout=5,
                )
        except requests.exceptions.ConnectionError as exc:
            self.logger.error("Connection to '%s' failed: %s", self.base_url, exc, exc_info=False)
            raise ConnectionError("Connection failed: run api_service")

        if response.status_code != 200:
            self.logger.error("Error while getting full sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")

        return response.json()

    def put_state(self):
        response = requests.put(
                f"{self.base_url}/calc/put_state",
                json=self.data,
                timeout=5,
            )

        if response.status_code == 400 and "detail" in response.json() and "Error" in response.json()["detail"]:
            return response.json()["detail"]['Error']

        if response.status_code != 200:
            self.logger.error("Error while getting full sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")


        data = response.json()
        self.data.update(data)
        self.logger.debug("CalcClient updated data %s", data)
        return ""

    def post(self, endpoint, json):
        response = requests.post(
                f"{self.base_url}/calc/{endpoint}",
                json=json,
                timeout=5,
            )

        if response.status_code == 400 and "detail" in response.json() and "Error" in response.json()["detail"]:
            return response.json()["detail"]['Error']

        if response.status_code != 200:
            self.logger.error("Error while getting full sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")


        data = response.json()
        self.data.update(data)
        self.logger.debug("CalcClient updated data %s", data)
        return ""

    def delete(self, endpoint, params=None):
        if params is None:
            params = {}

        response = requests.delete(
                f"{self.base_url}/calc/{endpoint}",
                timeout=5,
                params=params,
            )

        if response.status_code != 200:
            self.logger.error("Error while getting full sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")

        data = response.json()
        self.data.update(data)
        self.logger.debug("CalcClient updated data %s", data)

    # сеттеры
    def set_option(self, k, val):
        try:
            val = int(val)
            if val == self.data['options'][k]:
                self.logger.debug("Trying to set the same value %d to option %s", val, k)
                return True
            if val > 1:
                self.data['options'][k] = val
                error = self.put_state()
                if error:
                    self.logger.error("Error: %s", error)
                    return False
                self.logger.debug("Option %s set to value %d", k, val)
                return True
            self.logger.warning("Wrong value of %s. The value should be positive but %d is given", k, val)
        except TypeError:
            self.logger.warning("Wrong type of the value of %s. The type is %s, but int is needed", k, type(val))
        return False

    def set_new_expr(self, expr):
        error = self.post(endpoint="set_new_expr", json={"expr": expr})
        if error:
            self.logger.error("Error while setting new expression: %s", error)
        return error

    def set_se(self, expr) -> str:
        self.data['se'] = str(expr)
        self.put_state()
        error = self.put_state()
        if error:
            self.logger.error("Error: %s", error)
            return error
        self.logger.info("Sympy-expression set: se = %s", self.data['se'])
        return ""

    def set_expr(self, expr):
        self.data['expr'] = expr
        self.put_state() # no exception
        self.logger.info("New expression set: expr = %s", self.data['expr'])

    def set_sec(self, sec):
        self.data['sec'] = sec
        self.put_state() # no exception
        self.logger.info("Field 'sec' set: sec = %s", self.data['sec'])

    def set_value(self, var, expr):
        self.data['values'][var] = expr
        error = self.put_state()
        if error:
            self.logger.error("Error: %s", error)
            return error
        self.logger.info("Value of %s set to %s", var, self.data['value'][var])
        return ""

    def delete_all_values(self):
        self.delete(endpoint="delete_all_values")
        self.logger.info("Values of all variables deleted")

    def delete_value(self, var):
        self.delete(endpoint="delete_value", params={"var": var})
        self.logger.info("Values of variable %s deleted", var)

    # геттеры
    def get_nice(self):
        return self.data['nice']

    def get_expr(self):
        return self.data['expr']

    def get_sec(self):
        return self.data['sec']

    def get_se(self):
        return self.data['se']

    def get_var_names(self):
        return self.data['values'].keys()

    def get_options(self):
        return self.data['options']

    def get_help_text(self):
        return self.data['help_text']

    def get_explanations(self):
        return self.data['explanations']

    def get_variables(self, include_unset=True):
        out_json = self.get(endpoint="get_variables", params={"include_unset": include_unset})
        return out_json["current"], out_json['lines']

    def current_values(self):
        out_json = self.get(endpoint="current_values")
        return out_json["expr"]

    def evaluate(self):
        out_json = self.get(endpoint="evaluate")
        return out_json["expr"]

    def clear_all(self):
        self.delete(endpoint="clear_all")
        self.logger.info("State cleared")

    def save_state(self, session_id):
        response = requests.post(
                f"{self.base_url}/calc/save_state",
                params={"session_id": session_id},
                timeout=5,
            )

        if response.status_code != 200:
            self.logger.error("Error while getting full sate. Status code: %s. Details: %s", response.status_code, response.text)
            raise requests.HTTPError(f"Bad status code {response.status_code}: {response.json()}")


if __name__ == '__main__':
    logger = make_logger(name="console", file=True, console=True, level=logging.DEBUG)
    calc = CalcClient(logger=logger)
    calc.put_state()

File: cpp_client/src/calc_client.cpp
#include "calc_client.h"
#include "requests/json_dump.h"

CalcClient::CalcClient(const QString& baseUrl, int timeout, QObject* parent)
    : CalcHttpRequester(baseUrl, timeout, parent)  
{
    m_data = CalcHttpRequester::get("get_state", {{"full", true}});
    qInfo() << "CalcClient connected and loaded data:" << shortJson(m_data);
}

//Сеттеры с запросом действия

QString CalcClient::setNewExpr(const QString& expr)
{
    QString error = CalcHttpRequester::post("set_new_expr", {{"expr", expr}});
    if (error.isEmpty())
    {
        qInfo() << "New expression set: expr =" << m_data["expr"];
        qInfo() << "and parsed to sympy-expression: SE =" << m_data["se"];
    }
    else
    {
        qCritical() << "Error while setting new expression:" << error;
    }

    // QString key = "digits";
    // QJsonObject optObj = m_data["options"].toObject();
    // qDebug() << "1:" << optObj.contains(key);
    // qDebug() << "2:" << (optObj.value(key).type() == QJsonValue::Double);
    // qDebug() << "2:" << optObj[key].isDouble();
    // qDebug() << "3:" << optObj[key].toInt();
    // qDebug() << "m_data" << m_data;

    return error;
}

//Сеттеры с простыми запросами

void CalcClient::setExpr(const QString& expr)
{
    m_data["expr"] = expr;
    QString error = CalcHttpRequester::put_state(); // no exception
    assert (error.isEmpty());
    qInfo() << "Expression set: expr =" << m_data["expr"];
}

QString CalcClient::setSE(const QString& expr)
{
    m_data["se"] = expr;
    QString error = CalcHttpRequester::put_state();
    if (error.isEmpty())
    {
        qInfo() << "Sympy-expression set: SE =" << m_data["se"];
    }
    else
    {
        qCritical() << "Error while setting SE:" << error;
    }
    return error;

}

void CalcClient::setSEC(const QString& expr)
{
    m_data["sec"] = expr;
    QString error = CalcHttpRequester::put_state(); // no exception
    assert (error.isEmpty());
    qInfo() << "Sympy-expression 'calculated' set: SEC =" << m_data["sec"];
}

QString CalcClient::setValue(const QString& variable, const QString& value)
{
    assert (m_data["values"].isObject());
    QJsonObject valObj = m_data["values"].toObject();
    valObj[variable] = value;
    m_data["values"] = valObj;
    QString error = CalcHttpRequester::put_state();
    if (error.isEmpty())
    {
        qInfo() << "Value of" << variable << "set to" << m_data["values"].toObject()[variable];
    }
    else
    {
        qCritical() << "Error while setting value of a variable:" << error;
    }
    return error;

}

bool CalcClient::setOption(const QString& key, const int value)
{
    assert (m_data["options"].isObject());
    QJsonObject optObj = m_data["options"].toObject();
    if (optObj.contains(key) && optObj.value(key).type() == QJsonValue::Double && value == optObj[key].toInt())
    {
        qDebug() << "Trying to set the same value" << value << "to option" << key;
        return true;
    }

    if (key <= 0)
    {
        qWarning().nospace() << "Wrong value of " << key << ". The value should be positive, but " << value << " was given\n";
        return false;
    }

    optObj[key] = value;
    m_data["options"] = optObj;
    QString error = CalcHttpRequester::put_state();
    if (!error.isEmpty())
    {
        qCritical() << "Error while setting option:" << error;
        return false;
    }

    qInfo() << "Option" << key << "set to value" << value;

    return true;
}

//Чистые геттеры
QString CalcClient::getNice()
{
    return m_data["nice"].toString();
}

QString CalcClient::getExpr()
{
    return m_data["expr"].toString();
}

QString CalcClient::getSE()
{
    return m_data["se"].toString();
}

QString CalcClient::getSEC()
{
    return m_data["sec"].toString();
}

QStringList CalcClient::getVarNames()
{
    return m_data["values"].toObject().keys();
}

QJsonObject CalcClient::getOptions()
{
    return m_data["options"].toObject();
}

QString CalcClient::getHelpText()
{
    return m_data["help_text"].toString();
}

QJsonObject CalcClient::getExplanations()
{
    return m_data["explanations"].toObject();
}

//Геттеры с запросами
QJsonObject CalcClient::getVariables(bool includeUnset)
{
    return CalcHttpRequester::get("get_variables", {{"include_unset", includeUnset}});
}

QString CalcClient::getCurrentValues()
{
    QJsonObject outJson = CalcHttpRequester::get("current_values");
    return outJson["expr"].toString();
}

//Запросы действий

void CalcClient::delAllValues()
{
    CalcHttpRequester::del("delete_all_values");
    qInfo() << "Values of all variables deleted";
}

void CalcClient::delValue(const QString& variable)
{
    CalcHttpRequester::del("delete_value", {{"var", variable}});
    qInfo() << "Values of all variable" << variable << "deleted";
}

QString CalcClient::evaluate()
{
    QJsonObject outJson = CalcHttpRequester::get("evaluate");
    return outJson["expr"].toString();
    
}
void CalcClient::clearAll()
{
    CalcHttpRequester::del("clear_all");
    qInfo() << "State cleared";
}
